cmd/libs/go2idl/openapi-gen/generators/openapi.go

```diff
@@ -0,0 +1,544 @@
+	importLines := []string{}
+	for _, singleImport := range g.imports.ImportLines() {
+		if g.isOtherPackage(singleImport) {
+			importLines = append(importLines, singleImport)
+		}
+	}
+	return importLines
+}
+
+func argsFromType(t *types.Type) generator.Args {
+	return generator.Args{
+		"type":               t,
+		"OpenAPIDefinitions": types.Ref(openAPIGenPackagePath, "OpenAPIDefinitions"),
+		"OpenAPIDefinition":  types.Ref(openAPIGenPackagePath, "OpenAPIDefinition"),
+		"SpecSchemaType":     types.Ref(specPackagePath, "Schema"),
+	}
+}
+
+func (g *openAPIGen) Init(c *generator.Context, w io.Writer) error {
+	sw := generator.NewSnippetWriter(w, c, "$", "$")
+	sw.Do("func (_ $.type|raw$) OpenAPIDefinitions() *$.OpenAPIDefinitions|raw$ {\n", argsFromType(g.targetType))
+	sw.Do("return &$.OpenAPIDefinitions|raw${\n", argsFromType(nil))
+	return sw.Error()
+}
+
+func (g *openAPIGen) Finalize(c *generator.Context, w io.Writer) error {
+	sw := generator.NewSnippetWriter(w, c, "$", "$")
+	sw.Do("}\n}\n", nil)
+	return sw.Error()
+}
+
+func (g *openAPIGen) GenerateType(c *generator.Context, t *types.Type, w io.Writer) error {
+	glog.V(5).Infof("generating for type %v", t)
+	sw := generator.NewSnippetWriter(w, c, "$", "$")
+	err := newOpenAPITypeWriter(sw).generate(t)
+	if err != nil {
+		return err
+	}
+	return sw.Error()
+}
+
+func getJsonTags(m *types.Member) []string {
+	jsonTag := reflect.StructTag(m.Tags).Get("json")
+	if jsonTag == "" {
+		return []string{}
+	}
+	return strings.Split(jsonTag, ",")
+}
+
+func getReferableName(m *types.Member) string {
+	jsonTags := getJsonTags(m)
+	if len(jsonTags) > 0 {
+		if jsonTags[0] == "-" {
+			return ""
+		} else {
+			return jsonTags[0]
+		}
+	} else {
+		return m.Name
+	}
+}
+
+func isPropertyRequired(m *types.Member) bool {
+	jsonTags := getJsonTags(m)
+	// A property is required if it does not have omitempty value in its json tag
+	return len(jsonTags) < 2 || jsonTags[1] != "omitempty"
+}
+
+type openAPITypeWriter struct {
+	*generator.SnippetWriter
+	refTypes               map[string]*types.Type
+	GetDefinitionInterface *types.Type
+}
+
+func newOpenAPITypeWriter(sw *generator.SnippetWriter) openAPITypeWriter {
+	return openAPITypeWriter{
+		SnippetWriter: sw,
+		refTypes:      map[string]*types.Type{},
+	}
+}
+
+func hasOpenAPIDefinitionMethod(t *types.Type) bool {
+	for mn, mt := range t.Methods {
+		if mn != "OpenAPIDefinition" {
+			continue
+		}
+		if len(mt.Signature.Parameters) != 0 || len(mt.Signature.Results) != 1 {
+			return false
+		}
+		r := mt.Signature.Results[0]
+		if r.Name.Name != "OpenAPIDefinition" || r.Name.Package != openAPIGenPackagePath {
+			glog.Errorf("Err4 %v", t)
+			return false
+		}
+		return true
+	}
+	return false
+}
+
+func (g openAPITypeWriter) generate(t *types.Type) error {
+	// Only generate for struct type and ignore the rest
+	switch t.Kind {
+	case types.Struct:
+		args := argsFromType(t)
+		g.Do("\"$.$\": ", filepath.Base(t.Name.Package)+"."+t.Name.Name)
```

Comment:
So I guess this is fine. I guess you could turn that string concatenation into a function if you do it in more than one place, to be self documenting. Only if you feel like it.